### PR TITLE
Tempus 1073 file upload linking

### PR DESCRIPTION
**Pull request by RachanaGupta01**
Thank you for submitting a contribution to Tempus.

In order to streamline the review of the contribution please
ensure the following steps have been taken:

### For all changes:
- [x] Is there a Waffle ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with Tempus-XXXX where XXXX is the waffle number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically dev)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn clean install at the root Tempus folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] Have you added documentation for any UI or API related changes to the /docs folder

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: application/src/it/java/com/hashmapinc/server/controller/BaseUploadControllerTest.java [hunk 1]
```diff
@@ -59,77 +62,93 @@ public void beforeTest() throws Exception {
...
-        Assert.assertEquals("temp", retFileMetaData.getFileName());
+        Assert.assertEquals("file", retFileMetaData.getFileName());
+
     }
 
     @Test
     public void uploadFileReturnsNotFoundException() throws Exception {
         when(cloudStorageService.upload(any(), any(), any(), any())).thenReturn(false);
-        doPostFile("/api/file", multipartFile).andExpect(status().isNotFound());
+        Device device = createDevice();
```
Comment (anuj-akt) [inline application/src/it/java/com/hashmapinc/server/controller/BaseUploadControllerTest.java]: Extract device creation in before test method

File: application/src/main/java/com/hashmapinc/server/controller/UploadController.java [hunk 2]
```diff
@@ -45,9 +50,13 @@
...
     @PostMapping(value = "/file")
     @ResponseBody
-    public FileMetaData upload(@RequestParam("file") MultipartFile file) throws TempusException {
+    public FileMetaData upload(@RequestParam("file") MultipartFile file,
+                               @RequestParam Map<String, String> relatedEntityInfo) throws TempusException {
         try {
-            FileMetaData savedFileMetaData = checkNotNull(uploadService.uploadFile(file, getCurrentUser().getTenantId()));
+            String strRelatedEntityId = relatedEntityInfo.get("relatedEntityId");
+            String strRelatedEntityType = relatedEntityInfo.get("relatedEntityType");
+            EntityId relatedEntityId = createRelatedEntityId(strRelatedEntityId, strRelatedEntityType);//getRelatedEntityId(strRelatedEntityId);
```
Comment (anuj-akt) [inline application/src/main/java/com/hashmapinc/server/controller/UploadController.java]: clean comment

File: application/src/main/java/com/hashmapinc/server/service/upload/UploadService.java [hunk 3]
```diff
@@ -16,23 +16,22 @@
...
 
+import com.hashmapinc.server.common.data.id.EntityId;
 import com.hashmapinc.server.common.data.id.TenantId;
 import com.hashmapinc.server.common.data.upload.FileMetaData;
 import com.hashmapinc.server.common.data.upload.InputStreamWrapper;
 import com.hashmapinc.server.common.data.upload.StorageTypes;
+import com.hashmapinc.server.dao.filemetadata.FileMetaDataDao;
 import com.hashmapinc.server.dao.tenant.TenantService;
 import com.hashmapinc.server.service.CloudStorageServiceUtils;
 import com.hashmapinc.server.service.computation.CloudStorageService;
```
Comment (anuj-akt) [inline application/src/main/java/com/hashmapinc/server/service/upload/UploadService.java]: move it out of computation package

File: application/src/main/java/com/hashmapinc/server/service/upload/UploadService.java [hunk 4]
```diff
@@ -44,59 +43,92 @@
...
 
-    public void renameFile(String oldFileName, String newFileName, TenantId tenantId) throws Exception{
+    public void renameFile(String oldFileName, String newFileName, TenantId tenantId, EntityId entityId) throws Exception{
         String bucketName = CloudStorageServiceUtils.createBucketName(tenantService.findTenantById(tenantId));
-        String oldObjectUrl = CloudStorageServiceUtils.createObjectName(oldFileName, StorageTypes.FILES);
-        String newObjectUrl = CloudStorageServiceUtils.createObjectName(newFileName, StorageTypes.FILES);
-        if(cloudStorageService.copyFile(bucketName, oldObjectUrl, newObjectUrl)) {
+        String oldObjectUrl = CloudStorageServiceUtils.createObjectName(oldFileName, entityId, StorageTypes.FILES);
+        String newObjectUrl = CloudStorageServiceUtils.createObjectName(newFileName, entityId, StorageTypes.FILES);
+        if (!oldObjectUrl.contentEquals(newObjectUrl) && cloudStorageService.copyFile(bucketName, oldObjectUrl, newObjectUrl)) {
```
Comment (anuj-akt) [inline application/src/main/java/com/hashmapinc/server/service/upload/UploadService.java]: update filemetadata in this case

File: application/src/main/java/com/hashmapinc/server/service/upload/UploadService.java [hunk 5]
```diff
@@ -44,59 +43,92 @@
...
-            return new FileMetaData(fileNameWithoutExt, arrList2[arrList2.length - 1],
-                    item.lastModified().toInstant().getEpochSecond() * 1000,
-                    Precision.round(item.objectSize() / 1024.0, 4));
+    private FileMetaData createFileMetaData(MultipartFile file, TenantId tenantId, EntityId entityId) {
+        final String originalFilename = file.getOriginalFilename();
+        String[] arrList = originalFilename.split("\\.");
+        if (arrList.length >= 2) {
+            String fileNameWithoutExt = getFileNameWithoutExt(originalFilename);
+            return new FileMetaData(tenantId, entityId, entityId.getEntityType(), fileNameWithoutExt, arrList[arrList.length - 1],
+                    System.currentTimeMillis(),
```
Comment (anuj-akt) [inline application/src/main/java/com/hashmapinc/server/service/upload/UploadService.java]: refactor duplicate return metadta

**Review by anuj-akt — CHANGES_REQUESTED** [review]

**Review by jayeshkapadnis — APPROVED** [review]

**Review by anuj-akt — APPROVED** [review]

**Pull request merged**